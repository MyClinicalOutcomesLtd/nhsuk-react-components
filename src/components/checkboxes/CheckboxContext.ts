import { createContext, useContext } from 'react';

export interface ICheckboxContext {
  name: string;
  getBoxId: (reference: string) => string | undefined;
  setConditional: (boxReference: string, hasConditional: boolean) => void;
  leaseReference: () => string;
  unleaseReference: (reference: string) => void;
}

const CheckboxContext = createContext<ICheckboxContext>({
  /* eslint-disable @typescript-eslint/no-empty-function */
  name: '',
  getBoxId: () => undefined,
  setConditional: () => {},
  leaseReference: () => '',
  unleaseReference: () => {},
});

export const useCheckboxContext = (): ICheckboxContext => useContext(CheckboxContext);

export default CheckboxContext;
