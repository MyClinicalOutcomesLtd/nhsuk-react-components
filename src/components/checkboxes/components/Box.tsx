import classNames from 'classnames';
import React, { ComponentProps, HTMLProps, ReactNode, useEffect, useState } from 'react';
import LabelBlock from '../../../util/LabelBlock';
import Hint from '../../hint';
import Label from '../../label';
import { useCheckboxContext } from '../CheckboxContext';

type CheckboxesBoxProps = Omit<HTMLProps<HTMLInputElement>, 'label'> & {
  labelProps?: ComponentProps<typeof Label>;
  hint?: ReactNode;
  hintProps?: ComponentProps<typeof Hint>;
  conditional?: ReactNode;
  forceShowConditional?: boolean;
  conditionalWrapperProps?: HTMLProps<HTMLDivElement>;
};

const CheckboxesBox: React.FC<CheckboxesBoxProps> = React.forwardRef((props, ref) => {
  const boxCtx = useCheckboxContext();
  const [boxReference] = useState<string>(() => boxCtx.leaseReference());
  const [showConditional, setShowConditional] = useState(!!(props.checked || props.defaultChecked));

  const inputID = props.id || boxCtx.getBoxId(boxReference);

  useEffect(() => () => boxCtx.unleaseReference(boxReference), []);

  useEffect(() => {
    if (props.checked !== undefined) setShowConditional(props.checked);
  }, [props.checked]);

  useEffect(() => {
    boxCtx.setConditional(boxReference, Boolean(props.conditional));
    return () => boxCtx.setConditional(boxReference, false);
  }, [props.conditional]);

  const { children, ...inputProps } = props;

  return (
    <>
      <div className="nhsuk-checkboxes__item">
        <input
          {...inputProps}
          className={classNames('nhsuk-checkboxes__input', props.className)}
          onChange={(e) => {
            if (props.checked === undefined) setShowConditional(e.target.checked);
            if (props.onChange) props.onChange(e);
          }}
          id={inputID}
          ref={ref}
          name={inputProps.name || boxCtx.name}
        />
        <LabelBlock
          elementId={inputID}
          label={children}
          labelProps={{
            ...props.labelProps,
            className: classNames('nhsuk-checkboxes__label', props.labelProps?.className),
          }}
          hint={props.hint}
          hintProps={{
            ...props.hintProps,
            className: classNames('nhsuk-checkboxes__hint', props.hintProps?.className),
          }}
        />
      </div>
      {props.conditional && (showConditional || props.forceShowConditional) ? (
        <div
          {...props.conditionalWrapperProps}
          className={classNames(
            'nhsuk-radios__conditional',
            props.conditionalWrapperProps?.className,
          )}
          id={`${inputID}--conditional`}
        >
          {props.conditional}
        </div>
      ) : null}
    </>
  );
});

CheckboxesBox.displayName = 'Checkboxes.Box';
CheckboxesBox.defaultProps = {
  type: 'checkbox',
};

export default CheckboxesBox;
