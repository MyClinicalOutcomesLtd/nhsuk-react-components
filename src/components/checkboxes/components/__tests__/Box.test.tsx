import { mount } from 'enzyme';
import React from 'react';
import { act } from 'react-test-renderer';
import CheckboxContext from '../../CheckboxContext';
import Box from '../Box';

describe('Box', () => {
  it('renders successfully', () => {
    const context = {
      name: 'test',
      setConditional: jest.fn(),
      getBoxId: jest.fn().mockReturnValue('test-1'),
      leaseReference: jest.fn().mockReturnValue('test-reference'),
      unleaseReference: jest.fn(),
    };

    const wrapper = mount(
      <CheckboxContext.Provider value={context}>
        <Box>Test Box</Box>
      </CheckboxContext.Provider>,
    );

    expect(wrapper).toMatchSnapshot();

    expect(context.leaseReference).toHaveBeenCalledTimes(1);
    expect(context.getBoxId).toHaveBeenCalledTimes(1);
    expect(context.setConditional).toHaveBeenNthCalledWith(1, 'test-reference', false);
    expect(context.unleaseReference).not.toHaveBeenCalled();

    wrapper.unmount();

    expect(context.unleaseReference).toHaveBeenCalledTimes(1);
    expect(context.unleaseReference).toHaveBeenCalledWith('test-reference');
    expect(context.setConditional).toHaveBeenCalledTimes(2);
    expect(context.setConditional).toHaveBeenLastCalledWith('test-reference', false);
  });

  it('shows conditional content on click', () => {
    const context = {
      name: 'test',
      setConditional: jest.fn(),
      getBoxId: jest.fn().mockReturnValue('test-1'),
      leaseReference: jest.fn().mockReturnValue('test-reference'),
      unleaseReference: jest.fn(),
    };

    const wrapper = mount(
      <CheckboxContext.Provider value={context}>
        <Box conditional="Conditional Content">Test Box</Box>
      </CheckboxContext.Provider>,
    );

    expect(context.leaseReference).toHaveBeenCalledTimes(1);
    expect(context.getBoxId).toHaveBeenCalledTimes(1);
    expect(context.setConditional).toHaveBeenNthCalledWith(1, 'test-reference', true);
    expect(context.unleaseReference).not.toHaveBeenCalled();

    expect(wrapper.find('div.nhsuk-radios__conditional').exists()).toBe(false);

    const boxEl = wrapper.find('input');

    act(() => {
      boxEl.simulate('change', { target: { checked: true } });
    });

    const conditionalEl = wrapper.find('div.nhsuk-radios__conditional');
    expect(conditionalEl.exists()).toBe(true);
    expect(conditionalEl.prop('id')).toBe('test-1--conditional');
    expect(conditionalEl.text()).toBe('Conditional Content');

    wrapper.unmount();
  });

  it('shows conditional content if checked by default', () => {
    const context = {
      name: 'test',
      setConditional: jest.fn(),
      getBoxId: jest.fn().mockReturnValue('test-1'),
      leaseReference: jest.fn().mockReturnValue('test-reference'),
      unleaseReference: jest.fn(),
    };

    const wrapper = mount(
      <CheckboxContext.Provider value={context}>
        <Box conditional="Conditional Content" checked>
          Test Box
        </Box>
      </CheckboxContext.Provider>,
    );

    expect(context.leaseReference).toHaveBeenCalledTimes(1);
    expect(context.getBoxId).toHaveBeenCalledTimes(1);
    expect(context.setConditional).toHaveBeenNthCalledWith(1, 'test-reference', true);
    expect(context.unleaseReference).not.toHaveBeenCalled();

    const conditionalEl = wrapper.find('div.nhsuk-radios__conditional');
    expect(conditionalEl.exists()).toBe(true);
    expect(conditionalEl.prop('id')).toBe('test-1--conditional');
    expect(conditionalEl.text()).toBe('Conditional Content');

    wrapper.unmount();
  });

  it('does not render a label when not provided', () => {
    const wrapper = mount(<Box />);
    expect(wrapper.find('.nhsuk-label').exists()).toBe(false);
    wrapper.unmount();
  });

  it('does not change conditional state if checked value is provided', () => {
    const context = {
      name: 'test',
      setConditional: jest.fn(),
      getBoxId: jest.fn().mockReturnValue('test-1'),
      leaseReference: jest.fn().mockReturnValue('test-reference'),
      unleaseReference: jest.fn(),
    };

    const wrapper = mount(
      <CheckboxContext.Provider value={context}>
        <Box conditional="Conditional Content" checked={false}>
          Test Box
        </Box>
      </CheckboxContext.Provider>,
    );

    expect(context.leaseReference).toHaveBeenCalledTimes(1);
    expect(context.getBoxId).toHaveBeenCalledTimes(1);
    expect(context.setConditional).toHaveBeenNthCalledWith(1, 'test-reference', true);
    expect(context.unleaseReference).not.toHaveBeenCalled();

    expect(wrapper.find('div.nhsuk-radios__conditional').exists()).toBe(false);

    const boxEl = wrapper.find('input');

    act(() => {
      boxEl.simulate('change', { target: { checked: true } });
    });

    expect(wrapper.find('div.nhsuk-radios__conditional').exists()).toBe(false);

    wrapper.unmount();
  });

  it('handles event handlers', () => {
    const handleChange = jest.fn();

    const wrapper = mount(
      <Box conditional="Conditional Content" checked={false} onChange={handleChange}>
        Test Box
      </Box>,
    );

    const boxEl = wrapper.find('input');
    act(() => {
      boxEl.simulate('change', { target: { checked: true } });
    });

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange.mock.calls[0][0].target.checked).toBe(true);

    wrapper.unmount();
  });

  it('renders hint when provided', () => {
    const wrapper = mount(
      <Box conditional="Conditional Content" checked={false} hint="Test Hint">
        Test Box
      </Box>,
    );

    const hintEl = wrapper.find('div.nhsuk-hint');
    expect(hintEl.exists()).toBe(true);
    expect(hintEl.text()).toBe('Test Hint');

    wrapper.unmount();
  });
});
