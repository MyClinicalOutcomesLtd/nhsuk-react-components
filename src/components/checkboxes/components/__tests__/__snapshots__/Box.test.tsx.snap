// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders successfully 1`] = `
<Checkboxes.Box
  type="checkbox"
>
  <div
    className="nhsuk-checkboxes__item"
  >
    <input
      className="nhsuk-checkboxes__input"
      id="test-1"
      name="test"
      onChange={[Function]}
      type="checkbox"
    />
    <LabelBlock
      elementId="test-1"
      hintProps={
        Object {
          "className": "nhsuk-checkboxes__hint",
        }
      }
      label="Test Box"
      labelProps={
        Object {
          "className": "nhsuk-checkboxes__label",
        }
      }
    >
      <Label
        className="nhsuk-checkboxes__label"
      >
        <BaseLabel
          className="nhsuk-checkboxes__label"
        >
          <label
            className="nhsuk-label nhsuk-checkboxes__label"
          >
            Test Box
          </label>
        </BaseLabel>
      </Label>
    </LabelBlock>
  </div>
</Checkboxes.Box>
`;
